kong/db/dao/plugins/external.lua

```diff
@@ -0,0 +1,642 @@
+to be executed as a command line.  The output should be a JSON string that decodes
+as an array of objects, each defining the name, priority, version and schema of one
+plugin.
+
+    [{
+      "name": ... ,
+      "priority": ... ,
+      "version": ... ,
+      "schema": ... ,
+      "phases": [ phase_names ... ],
+    },
+    {
+      ...
+    },
+    ...
+    ]
+
+This array should describe all plugins currently available through this server,
+no matter if actually enabled in Kong's configuration or not.
+
+--]]
+
+
+local function register_plugin_info(server_def, plugin_info)
+  if _plugin_infos[plugin_info.Name] then
+    kong.log.err(string.format("Duplicate plugin name [%s] by %s and %s",
+      plugin_info.Name, _plugin_infos[plugin_info.Name].server_def.name, server_def.name))
+    return
+  end
+
+  _plugin_infos[plugin_info.Name] = {
+    server_def = server_def,
+    --rpc = server_def.rpc,
+    name = plugin_info.Name,
+    PRIORITY = plugin_info.Priority,
+    VERSION = plugin_info.Version,
+    schema = plugin_info.Schema,
+    phases = plugin_info.Phases,
+  }
+end
+
+local function ask_info(server_def)
+  if not server_def.query_command then
+    kong.log.info(string.format("No info query for %s", server_def.name))
+    return
+  end
+
+  local fd, err = io.popen(server_def.query_command)
+  if not fd then
+    local msg = string.format("loading plugins info from [%s]:\n", server_def.name)
+    kong.log.err(msg, err)
+    return
+  end
+
+  local infos_dump = fd:read("*a")
+  fd:close()
+  local infos = cjson_decode(infos_dump)
+  if type(infos) ~= "table" then
+    error(string.format("Not a plugin info table: \n%s\n%s",
+        server_def.query_command, infos_dump))
+    return
+  end
+
+  for _, plugin_info in ipairs(infos) do
+    register_plugin_info(server_def, plugin_info)
+  end
+end
+
+local function load_all_infos()
+  if not _plugin_infos then
+    _plugin_infos = {}
+
+    for _, server_def in ipairs(get_server_defs()) do
+      ask_info(server_def)
+    end
+  end
+
+  return _plugin_infos
+end
+
+
+local loaded_plugins = {}
+
+local function get_plugin(plugin_name)
+  if not loaded_plugins[plugin_name] then
+    local plugin = load_all_infos()[plugin_name]
+    loaded_plugins[plugin_name] = build_phases(plugin)
+  end
+
+  return loaded_plugins[plugin_name]
+end
+
+function external_plugins.load_plugin(plugin_name)
+  local plugin = get_plugin(plugin_name)
+  if plugin and plugin.PRIORITY then
+    return true, plugin
+  end
+
+  return false, "no plugin found"
+end
+
+function external_plugins.load_schema(plugin_name)
+  local plugin = get_plugin(plugin_name)
+  if plugin and plugin.PRIORITY then
+    return true, plugin.schema
+  end
+
+  return false, "no plugin found"
+end
+
+
+--[[
+
+Process management
+
+Servers that specify an `.exec` field are launched and managed by Kong.
+This is an attempt to duplicate the smallest reasonable subset of systemd.
```

Comment:
"Manage" is defined in an indirect way using systemd, which people reading this might not be familiar with. I the word "Manage" does not support its own weight here. I think we can explain what's happening without mentioning "Manage" or systemd at all. Something like (please correct, and note that I am still unsure about the difference between `exec` and `start_command`):

> If a Plugin Server has an `.exec` field, then Kong will monitor its processes, relaunching them if they die:
> * Restarting it using its `.exec`, `arguments` and environment fields.
> * Stdout and stderr are joined and logged

Reply:
I would also explicitly mention what happens if `.exec` isn't there. Example:

> If a Plugin server has no `.exec` field, it is assumed that this monitorization happens somewhere else. Kong will not try to restart its process, nor log anything if it dies.

Reply:
there's no `.exec` anymore (nor `.arguments` or `.environment`)  they all were part of the yaml file.  i have to rewrite these comments completely.  not keeping double negatives isn't bad advice. :wink: